diff --git a/OXInterface/OXInterface.py b/OXInterface/OXInterface.py
index ed2d40f..b880b04 100644
--- a/OXInterface/OXInterface.py
+++ b/OXInterface/OXInterface.py
@@ -22,7 +22,7 @@ from datetime import datetime
 from functools import lru_cache
 from typing import List, Dict, Any, NoReturn
 from onnx.numpy_helper import from_array, to_array
-from onnx.onnx_ml_pb2 import TensorProto, ValueInfoProto, AttributeProto
+from onnx import TensorProto, ValueInfoProto, AttributeProto
 from onnx.helper import make_attribute, make_node, make_graph, make_model
 from skl2onnx.helpers.onnx_helper import enumerate_model_node_outputs, select_model_inputs_outputs, save_onnx_model
 
