--- FLAVR/pytorch_msssim/__init__.py	2022-10-27 15:42:13.493376660 +0800
+++ __init__.py	2022-10-27 17:41:28.445464323 +0800
@@ -20,7 +20,7 @@
     _1D_window = gaussian(window_size, 1.5).unsqueeze(1)
     _2D_window = _1D_window.mm(_1D_window.t())
     _3D_window = _2D_window.unsqueeze(2) @ (_1D_window.t())
-    window = _3D_window.expand(1, channel, window_size, window_size, window_size).contiguous().cuda()
+    window = _3D_window.expand(1, channel, window_size, window_size, window_size).contiguous()
     return window
 
 
