diff --git a/PyTorch/SpeechSynthesis/FastPitch/fastpitch/data_function.py b/PyTorch/SpeechSynthesis/FastPitch/fastpitch/data_function.py
index 8e64e97d..062ef6a8 100644
--- a/PyTorch/SpeechSynthesis/FastPitch/fastpitch/data_function.py
+++ b/PyTorch/SpeechSynthesis/FastPitch/fastpitch/data_function.py
@@ -51,7 +51,8 @@ class BetaBinomialInterpolator:
     def __init__(self, round_mel_len_to=100, round_text_len_to=20):
         self.round_mel_len_to = round_mel_len_to
         self.round_text_len_to = round_text_len_to
-        self.bank = functools.lru_cache(beta_binomial_prior_distribution)
+        f = functools.lru_cache(maxsize=128)
+        self.bank = f(beta_binomial_prior_distribution)
 
     def round(self, val, to):
         return max(1, int(np.round((val + 1) / to))) * to
