diff --git a/export.py b/export.py
index 15e92a7..f381e0e 100644
--- a/export.py
+++ b/export.py
@@ -542,7 +542,7 @@ def parse_opt():
     parser.add_argument('--iou-thres', type=float, default=0.45, help='TF.js NMS: IoU threshold')
     parser.add_argument('--conf-thres', type=float, default=0.25, help='TF.js NMS: confidence threshold')
     parser.add_argument('--include', nargs='+',
-                        default=['torchscript', 'onnx'],
+                        default=['onnx'],
                         help='torchscript, onnx, openvino, engine, coreml, saved_model, pb, tflite, edgetpu, tfjs')
     opt = parser.parse_args()
     print_args(FILE.stem, opt)
diff --git a/models/yolo.py b/models/yolo.py
index f659a04..a2edd82 100644
--- a/models/yolo.py
+++ b/models/yolo.py
@@ -48,8 +48,11 @@ class Detect(nn.Module):
 
     def forward(self, x):
         z = []  # inference output
+        self.training = True
         for i in range(self.nl):
             x[i] = self.m[i](x[i])  # conv
+            if torch.onnx.is_in_onnx_export():
+                continue
             bs, _, ny, nx = x[i].shape  # x(bs,255,20,20) to x(bs,3,20,20,85)
             x[i] = x[i].view(bs, self.na, self.no, ny, nx).permute(0, 1, 3, 4, 2).contiguous()
 
