--- ./models/convnext.py	2022-09-05 18:18:52.922170456 -0900
+++ ./models/convnext.py	2022-09-05 18:22:04.234163614 -0900
@@ -109,7 +109,7 @@
         for i in range(4):
             x = self.downsample_layers[i](x)
             x = self.stages[i](x)
-        return self.norm(x.mean([-2, -1])) # global average pooling, (N, C, H, W) -> (N, C)
+        return self.norm(x.mean([2, 3])) # global average pooling, (N, C, H, W) -> (N, C)
 
     def forward(self, x):
         x = self.forward_features(x)
