--- Chinese-CLIP/cn_clip/deploy/pytorch_to_onnx.py	2023-09-04 10:01:33.630208000 +0800
+++ CLIP/pytorch_to_onnx.py	2023-09-04 16:20:55.346517000 +0800
@@ -126,10 +126,6 @@
         text_fp16_onnx_model = convert_float_to_float16(text_fp32_onnx_model, keep_io_types=True, disable_shape_infer=True)
         save_model(text_fp16_onnx_model,
                     text_fp16_onnx_path,
-                    location="{}.extra_file".format(os.path.split(text_fp16_onnx_path)[1]),
-                    save_as_external_data=True,
-                    all_tensors_to_one_file=True,
-                    size_threshold=1024,
                     convert_attribute=True)
 
     if args.convert_vision:
@@ -155,10 +151,6 @@
         vision_fp16_onnx_model = convert_float_to_float16(vision_fp32_onnx_model, keep_io_types=True, disable_shape_infer=True)
         save_model(vision_fp16_onnx_model,
                     vision_fp16_onnx_path,
-                    location="{}.extra_file".format(os.path.split(vision_fp16_onnx_path)[1]),
-                    save_as_external_data=True,
-                    all_tensors_to_one_file=True,
-                    size_threshold=1024,
                     convert_attribute=True)
 
     print("Finished PyTorch to ONNX conversion...")
